### ISSUE TITLE
[FEATURE] Make SOF with Zephyr run on i.MX93's A55 core with Jailhouse

### ISSUE THREAD

**Issue by LaurentiuM1234** [thesofproject/sof#7192]
**PURPOSE**
I've decided to open this issue in order to track the current progress and interact with the community. Our plan is to make SOF with Zephyr run on i.MX93's cortex-A55 core with the help of Jailhouse. Jailhouse will be used to split the hardware resources between Linux and Zephyr which will be running independently on different cores.

Feedback and suggestions would be very much appreciated.

**ISSUES**

1. Currently, using `scripts/xtensa-build-zephyr`, I have managed to obtain the `zephyr.elf`. The script fails when trying to run `smex` in order to build the `.ldc` file. My current understanding is that `smex` isn't able to properly parse the `zephyr.elf` file because of the fact that there's currently no support for parsing `ELF-64` files. My question regarding this would be: would it be a good idea to add support in `smex` for parsing such files? Also, would it be necessary since, for now, my plan for now is to use Zephyr's logging system (no trace, the logs will be printed via the UART)?
2. Regarding `zephyr.elf`'s sections: are there any restrictions imposed by SOF (e.g: alignment, order)? For now, I was able to get away by simply adding the `.static_uuids`, `.static_log` and `.trace_ctx` to the already present sections added for ARM64 platforms on Zephyr.
3. Would going through `smex` and the signing process still be required if the plan is to load the firmware image through Jailhouse? Please correct me if I'm wrong but my current understanding is that the .ri file resulting from the SOF build process contains some sort of header which is "stripped" during the load process in the linux kernel. Because of this, since Jailhouse loads the firmware image as-is, the A55 core would probably not be able to work properly. Would it be acceptable to bypass the .ri file creation process and simply utilize the generated `zephyr.bin`? (which can be generated by enabling `CONFIG_BUILD_OUTPUT_BIN`)

**RELATED PRs**
TODO

CC: @dbaluta @iuliana-prodan @paulstelian97 @lgirdwood

**Comment by lgirdwood** [comment]
@LaurentiuM1234 what is jailhouse ? any links ?
Adding @nashif @mwasko @andyross

**Comment by LaurentiuM1234** [comment]
@lgirdwood Sure, for a more detailed overview please see [1] and [2]. For a more general overview please see [3]. Also, please see [4] for github repo.

[1]: https://lwn.net/Articles/578295/
[2]: https://lwn.net/Articles/578852/
[3]: http://events17.linuxfoundation.org/sites/events/files/slides/ELCE2016-Jailhouse-Tutorial.pdf
[4]: https://github.com/siemens/jailhouse

TL; DR: Jailhouse is a hypervisor used to partition the system's resources. In our case, we'll be using Jailhouse to split the A55 cores between Linux and Zephyr, meaning they will run independently on different cores.

**Comment by lgirdwood** [comment]
@LaurentiuM1234 thanks, had a read through of the jailhouse docs. Looks like a great feature.  Re questions above.
1) Re the Zephyr logging / dictionary support - I think this is mostly done now, although there could be minor work still WIP. Best to check with @kv2019i and @marc-hb 
2) There is some WIP to update the Zephyr sections, but I dont think we have any ordering restrictions in general (maybe an xtensa thing for literals only). @andyross would know.
3) No need to use smex/rimage, I think you could just use teh default Zephyr Jailhouse methods for image building and running the code. i.e. use elf or bin files if that is what jailhouse needs.

**Comment by kv2019i** [comment]
@LaurentiuM1234  Ack, no need to use smex/rimage. Zephyr dictionary support can be used and we have tested it to work. Only reason why we still don't have dictionary enabled by default for Intel targets is that we have many platforms stuck to old compiler versions and Zephyr dictionary require a somewhat modern compiler. We just switched one Intel platform to clang toolchain and that will allow us to us dictionaries. With that, no need for smex anymore.

**Comment by andyross** [comment]
Regarding linker behavior: that gets pretty hairy internal to zephyr, but for the most part apps don't need to care.  SOF has a few special sections (e.g. rimage manifest data) that you'll need to implement in the Zephyr SOC layer (probably just clone from the intel_adsp linker scripts, though it's possible your requirements are mild if you aren't using rimage itself).  Definitely ping me if you have trouble.

**Comment by marc-hb** [comment]
`west sign` should be removed from `xtensa-build-zephyr.py` in the next couple months, please look at:
- https://github.com/zephyrproject-rtos/zephyr/pull/54700

It should soon be part of `west build` instead for intel_adsp. For i.MX it will be all up to you!

**Comment by marc-hb** [comment]
> @LaurentiuM1234 Ack, no need to use smex/rimage.

`rimage` is definitely used by i.MX, thanks @dbaluta for explaining that to me. `rimage` is not just signing but also "stitching"

For smex, you could easily add some new `SMEX_BUILD = False` property after #7255 is merged. Or you could just add ELF64 support to smex, the smex code is surprisingly small and it looks like it has some ELF64 headers already.

**Comment by marc-hb** [comment]
> Would it be acceptable to bypass the .ri file creation process and simply utilize the generated zephyr.bin? (which can be generated by enabling CONFIG_BUILD_OUTPUT_BIN)

You don't need to disable `rimage` to use `zephyr.bin`.

If you need neither rimage nor smex you probably don't want to use `xtensa-build-zephyr.py` at all. Look at the `west` commands invoked by `xtensa-build-zephyr.py` and use `west` directly.

**Comment by LaurentiuM1234** [comment]
Thank you very much for all the useful comments!



> > Would it be acceptable to bypass the .ri file creation process and simply utilize the generated zephyr.bin? (which can be generated by enabling CONFIG_BUILD_OUTPUT_BIN)
> 
> You don't need to disable `rimage` to use `zephyr.bin`.
> 
> If you need neither rimage nor smex you probably don't want to use `xtensa-build-zephyr.py` at all. Look at the `west` commands invoked by `xtensa-build-zephyr.py` for "inspiration", then invoke `west` directly.
> 
> Everything is possible :-)

Hm, for now I've used `xtensa-build-zephyr.py` for the sake of convenience since it also takes care of cloning the Zephyr repo. Although `smex` and `rimage` are not required for i.MX93 with Jailhouse (at least as far as I'm aware right now) I think  we can still use `xtensa-build-zephyr.py` with something like this:


> For smex, you could easily add some new `SMEX_BUILD = False` property after #7255 is merged. Or you could just add ELF64 support to smex, the smex code is surprisingly small and it looks like it has some ELF64 headers already.

and maybe `RIMAGE_BUILD = False` just to be somewhat consistent with the other platforms and have the convenience of `xtensa-build-zephyr.py` also cloning Zephyr for us (although this can be obtained by calling the script with the `-u` argument and then manually doing the `west` commands as @marc-hb suggested).

An alternative to this would be to have something like `sof-build-jailhouse.py` that would create the `zephyr.bin` and also do the Jailhouse setup required (in case there's more people interested in adding support for their platforms to run SOF with Jailhouse). But at this point this is just an idea which I haven't tested yet so we'll have to see if it's worth introducing such a script. 

**PROGRESS UPDATE**

On Zephyr and SOF sides of things everything seems to be working fine, managed to get some logs on the console. Currently, I'm working on the Linux side of things. Hopefully, will soon have something decent enough to submit a draft PR.

**Comment by marc-hb** [comment]
> Hm, for now I've used xtensa-build-zephyr.py for the sake of convenience since it also takes care of cloning the Zephyr repo.

Cloning is just two `west` commands: `west init` and `west update`, look at the `xtensa-build-zephyr.py` logs. If that's too inconvenient then `west` is not doing it's number 1 job!?

Don't get me wrong: `xtensa-build-zephyr.py` does a number of things that `west` does not and it has been very useful to tame the typical configuration divergence between developers and CIs. But at the end of the day it's still a wrapper on top of wrapper (west) on top of a wrapper (cmake) on top of wrapper (ninja) = best avoided if you don't need most of it. In the middle/long term we want to **reduce** the roles of `xtensa-build-zephyr.py`, not increase them or its complexity.

Current `xtensa-build-zephyr.py` jobs:
- Cloning: `west`'s number 1 job.
- Simplifying and centralizing the awkard XTENSA_ configuration - except it never did that for i.MX. I think @aborisovich is looking at CMake-based, alternative ways to do that in the future.
- A super awkward -i IPC4 overlay option that should be deleted and replaced by the more generic `-o any_overlay.conf`
- smex and sof-logger: becoming obsolete for Zephyr
- rimage: you say don't need it and I'm trying to move it to `west sign`, hopefully soon out of `xtensa-build-zephyr.py`.
- Laying out the `/lib/firmware/` and `sof-info` file structures and checksumming the outputs - do you care about these? They haven't been mentioned so far.

I think `SMEX_BUILD = False` and `RIMAGE_BUILD = False` options would be small enough changes hence OK but it would be a pity to add this extra code and small step towards combinatorial explosion if all you use the script for is `west` cloning...

Less is more!

Also like this one: "A little copying (= your own, tiny west script) is better than a little dependency (using only 10% of the xtensa-build-zephyr.py script)"

**Comment by lgirdwood** [comment]
@mmaka1 @kv2019i @plbossart fyi - good technical description here.

**Comment by LaurentiuM1234** [comment]
**STATUS UPDATE**
https://github.com/thesofproject/sof/pull/7425 has been broken down into the following PRs:

1. https://github.com/thesofproject/sof/pull/7458
2. https://github.com/thesofproject/sof/pull/7459
3. https://github.com/thesofproject/sof/pull/7460
4. https://github.com/thesofproject/sof/pull/7461
5. https://github.com/thesofproject/sof/pull/7425
6. https://github.com/thesofproject/sof/pull/7462
7. https://github.com/thesofproject/sof/pull/7463
8. https://github.com/thesofproject/sof/pull/7464
9. https://github.com/thesofproject/sof/pull/7465
10. https://github.com/thesofproject/sof/pull/7466
11. https://github.com/thesofproject/sof/pull/7467

This should hopefully make the reviewing process much easier. The below diagram depicts the dependency graph of the PRs:
![Screenshot from 2023-04-19 12-58-08](https://user-images.githubusercontent.com/26043217/233038756-83ac4705-0c8e-49de-a383-3c3c89e62c85.png)

**Note**: x -> y means PR x depends on PR y.

Zephyr PR https://github.com/zephyrproject-rtos/zephyr/pull/56713#event-9043543107 has already been merged. ATM working on Linux patches.

**Comment by marc-hb** [comment]
> Re the Zephyr logging / dictionary support - I think this is mostly done now, although there could be minor work still WIP. Best to check with @kv2019i and @marc-hb

Latest dictionary news from @kv2019i:
- https://github.com/thesofproject/sof/issues/7703


> `west sign` should be removed from `xtensa-build-zephyr.py` in the next couple months, `west sign` should soon be part of `west build` instead.

This is now done and the corresponding Zephyr documentation updates have just been merged:
- https://github.com/zephyrproject-rtos/zephyr/pull/58215/files
- For better NXP flexibility please review WEST_SIGN_OPTS one-line change https://github.com/zephyrproject-rtos/zephyr/pull/58804

This makes it much easier to skip `rimage` entirely.

I'm curious whether you're still using `xtensa-build-zephyr.py`? Only for cloning/initialisation convenience or also for building? Again: not using this script at all would be perfectly fine; it should always stay optional.


>  there's currently no support for parsing ELF-64 files
> [...]
>  Warnings during build process such as:
 `warning: format '%d' expects argument of type 'int', but argument 2 has type 'size_t' {aka 'long unsigned int'} [-Wformat=]`
and
`warning: taking address of packed member of 'struct coherent' may result in an unaligned pointer value [-Waddress-of-packed-member]`

I just saw that the A55 has a ILP32 mode, I'm curious why you didn't try that first?4GB looks like a lot of RAM for an audio DSP :-)

Lack of hybrid 32/64 support in Jailhouse? Or because of the way you implemented IPC? 

I'm aware 64 bits support is now old news in Zephyr but I doubt anyone else has been running SOF in LP64 mode yet... thanks for charting this new territory for the rest of us! :-)


> For smex, you could easily add some new SMEX_BUILD = False property after #7255 is merged.

Actually, that was probably a bad idea because I guess some users may want SMEX and other users not for the _same_  `xtensa-build-zephyr.py` configuration. So a new `--smex=False` CLI option would probably be better.

**Comment by LaurentiuM1234** [comment]
> I'm curious whether you're still using `xtensa-build-zephyr.py`? Only for cloning/initialisation convenience or also for building? Again: not using this script at all would be perfectly fine; it should always stay optional.

Currently, I'm using a modified version of `xtensa-build-zephyr.py` for cloning + building (by modified I mean it's basically the same thing with some lines of code commented out). Lately, I haven't really had the time to look more into the build process and decide/find a proper way to build the binary :(.

> I just saw that the A55 has a ILP32 mode, I'm curious why you didn't try that first?4GB looks like a lot of RAM for an audio DSP :-)

> Lack of hybrid 32/64 support in Jailhouse? Or because of the way you implemented IPC?

Frankly, I wasn't aware of the existence of this ABI. Thanks for pointing it out ! Might be worth investigating it a bit for the sake of curiosity at least.

**Comment by marc-hb** [comment]
I tried these but quickly realized there's a LOT missing in SOF for them:

```
west build -p --board mimx8mm_evk_a53 modules/audio/sof/app
west build -p --board mimx93_evk_a55  modules/audio/sof/app
```

What Zephyr `--board` have you been using?

**Comment by LaurentiuM1234** [comment]
> west build -p --board mimx8mm_evk_a53 modules/audio/sof/app

This shouldn't work as i.MX8MM with ARM64 is not supported in SOF.

> west build -p --board mimx93_evk_a55  modules/audio/sof/app

The name of the board to use is `mimx93_evk_a55_sof`.

Btw, were these command issued from Zephyr? If so, the reason the build fails is because the sof module is not updated, hence it doesn't contain the i.MX93 support.

If the command was issued from SOF side, then please make sure to update `west.yml` to also fetch the NXP HAL. Currently, i.MX93 support depends on it.

If any of the above doesn't work would it be possible to share the build errors?

**Comment by dbaluta** [comment]
> If the command was issued from SOF side, then please make sure to update `west.yml` to also fetch the NXP HAL. Currently, i.MX93 support depends on it.

I wonder how do we conditionally pull in a HAL for a specific platform family? (e.g IMX)

Currently, this works in SOF source tree root:

```
--- a/west.yml
+++ b/west.yml
@@ -60,6 +60,7 @@ manifest:
           - mipi-sys-t
           - lz4
           - tinycrypt
+          - hal_nxp
```

But there must be another way - to avoid unnecessary pulling `hal_nxp` for non-nxp platforms.

**Comment by marc-hb** [comment]
> ... and add a Github Action for it to make sure the codebase stays 64bits clean.

Done:
- #7772 

> If the command was issued from SOF side

I assumed you meant: "when using the sof/west.yml manifest".

> The name of the board to use is mimx93_evk_a55_sof. [...]  please make sure to update west.yml to also fetch the NXP HAL.

Thanks, that recipe finally worked out of the box! I tried a number of other configurations (from both manifest "sides") but none had worked so far.


> I was planning to eventually have a look at the warnings and find a way to fix them.

Looks like warnings sometimes point at real, hard problems! At least one in this LP64 case: the old sof-logger is still hardcoded to four 32bits parameters. I don't see how to solve this.

> But there must be another way - to avoid unnecessary pulling hal_nxp for non-nxp platforms.

It's very complicated actually and has been changing in west very recently:
- https://github.com/zephyrproject-rtos/zephyr/issues/57198
- https://github.com/zephyrproject-rtos/west/issues?q=+label%3A%22Partial+imports%22+

In #7772 I kept it simple and I just added `hal_nxp`. It's not that big and the `--fetch-opt=--filter=tree:0` git optimization makes it smaller and faster in CI. This seems to add about 20s in https://github.com/thesofproject/sof/actions/runs/5215791991, I think it's worth it.

Once the `west` dust settles down we can revisit and try to optimize.
